packages/driver/src/cypress/test_attempts.ts

```diff
@@ -0,0 +1,83 @@
+export const testAttemptFinished = ({ runner, test, tests, err, constants, next }) => {
+  if (!err && !test.hasAttemptPassed) {
+    return false
+  }
+
+  const { STATE_PASSED, STATE_FAILED, HOOK_TYPE_AFTER_EACH } = constants
+
+  if (test.hasAttemptPassed) {
+    // Currently, to get passing attempts to rerun in mocha,
+    // we signal to mocha that we MIGHT need to retry a passed test attempt.
+    // If the test is run and there are no errors present, we assume a
+    // passed test attempt(set in ./driver/src/cypress/runner.ts)
+    test.state = STATE_PASSED
+  } else {
+    // Otherwise, we can assume the test attempt failed as 'err' would have to be present here.
+    test.state = STATE_FAILED
+  }
+
+  // Evaluate if the test should continue based on 'calculateTestStatus'.
+  // This is a custom method added by Cypress in ./driver/src/cypress/mocha.ts
+  let testStatusInfo = test.calculateTestStatus()
+
+  if (!testStatusInfo.shouldAttemptsContinue) {
+    // If the test has met the exit condition, we need to grab the metadata from
+    // 'calculateTestStatus' in order to display and interpret the test outerStatus correctly.
+    test._cypressTestStatusInfo = testStatusInfo
+
+    if (testStatusInfo.attempts > 1) {
+      // If the test has been run AT LEAST twice (i.e. we are retrying), and the exit condition is met,
+      // modify mocha '_retries' to be the max retries made in order to possibly short circuit a suite
+      // if a hook has failed on every attempt (which we may not know at this stage of the test run).
+
+      // We will need the original retries to 'reset' the possible retries
+      // if the test attempt passes and fits the exit condition, BUT an 'afterEach' hook fails.
+      // In this case, we need to know how many retries we can reapply to satisfy the config.
+      test._maxRetries = test._retries
+      test._retries = test._currentRetry
+    }
+  }
+
+  let retry = test.currentRetry()
+
+  // TODO: refactor this
+  // requeue the test if we have retries and haven't satisfied our retry configuration.
+  if (retry < test.retries() && testStatusInfo.shouldAttemptsContinue) {
```

Comment:
@MuazOthman pointing out that this is a bug bc retries may not be on but you could still have burn-in enabled, this isn't a problem in this PR, it'll be a problem in the burning PR.